allennlp/training/metrics/rouge.py

```diff
@@ -32,43 +32,131 @@ class ROUGE(Metric):
     def __init__(self, ngram_size: int = 2, exclude_indices: Set[int] = None,) -> None:
         self._ngram_size = ngram_size
         self._exclude_indices = exclude_indices or set()
-        self._recall_matches: Dict[int, int] = Counter()
-        self._recall_totals: Dict[int, int] = Counter()
+
+        self._total_rouge_n_recalls: Dict[int, float] = defaultdict(float)
+        self._total_rouge_n_precisions: Dict[int, float] = defaultdict(float)
+        self._total_rouge_n_f1s: Dict[int, float] = defaultdict(float)
+
+        self._total_rouge_l_f1 = 0.0
+
+        self._total_sequence_count = 0
 
     @overrides
     def reset(self) -> None:
-        self._recall_matches = Counter()
-        self._recall_totals = Counter()
+        self._total_rouge_n_recalls = defaultdict(float)
+        self._total_rouge_n_precisions = defaultdict(float)
+        self._total_rouge_n_f1s = defaultdict(float)
+
+        self._total_rouge_l_f1 = 0.0
+
+        self._total_sequence_count = 0
+
+    def _longest_common_subsequence(self, seq_1: torch.LongTensor, seq_2: torch.LongTensor):
+        """
+        Computes the longest common subsequences between `seq_1` and `seq_2`, ignoring `self._exclude_indices`.
+        """
+        m = len(seq_1)
+        n = len(seq_2)
+
+        # Slightly lower memory usage by iterating over the longer sequence in outer loop
+        # and storing previous lcs for the shorter sequence
+        if m < n:
+            seq_1, seq_2 = seq_2, seq_1
+            m, n = n, m
+
+        prev_lcs = torch.zeros(n + 1, dtype=torch.long)
+
+        for i in range(m - 1, -1, -1):
+            # Make sure we don't count special tokens as part of the subsequences
+            if seq_1[i].item() in self._exclude_indices:
+                continue
+
+            cur_lcs = torch.zeros_like(prev_lcs)
+            for j in range(n - 1, -1, -1):
+                if seq_1[i] == seq_2[j]:
```

Comment:
No, since the seq1 check assures that we won't have equality there if `seq2` contains one of the ids `_exclude_indices`. For the dynamic programming to work properly, we need still need to update the table `cur_lcs` at index `j`, even if `seq_2[j]` is an excluded id.